### PR TITLE
GH-25: TokenHolder unit tests.

### LINKED ISSUE #25: TokenHolder unit tests.

### PR DESCRIPTION
**Pull request by pgev**
fixes #25

### CODE REVIEW

File: test/test_lib/utils.js [hunk 1]
```diff
@@ -60,3 +61,35 @@ module.exports.expectRevert = async (
...
+            return reject(err);
+        }
+
+        const newBlockHash = web3.eth.getBlock('latest').hash;
+
+        return resolve(newBlockHash);
+    });
+});
+
+module.exports.advanceBlock = () => new Promise((resolve, reject) => {
```
[the remark this answers is not in the document]
Comment (pgev) [reply test/test_lib/utils.js]: Somehow compiler did not complain :). Nice catch. Done!

File: test/token_holder/submit_authorize_session.js [hunk 2]
```diff
@@ -0,0 +1,556 @@
...
+                    spendingLimit,
+                    expirationHeight,
+                    {
+                        from: registeredWallet0,
+                    },
+                ),
+                'Ephemeral key can not be null.',
+            );
+        });
+        it('Submitted ephemeral key is authorized.', async () => {
```
[the remark this answers is not in the document]
Comment (pgev) [reply test/token_holder/submit_authorize_session.js]: Done.

File: test/token_holder/revoke_session.js [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+
+            const keyData = await tokenHolder.ephemeralKeys.call(ephemeralKey);
+
+            assert.isOk(
+                // TokenHolder.AuthorizationStatus.AUTHORIZED == 1
+                keyData.status.eqn(1),
+            );
+
+            await utils.expectRevert(
+                tokenHolder.submitAuthorizeSession(
```
[the remark this answers is not in the document]
Comment (pgev) [reply test/token_holder/revoke_session.js]: The changes of revokeSession was not merged once I was writing this test cases :). I will merge with develop and apply appropriate changes here. Thx.

File: test/token_holder/revoke_session.js [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+
+const BN = require('bn.js');
+const web3 = require('../test_lib/web3.js');
+const utils = require('../test_lib/utils.js');
+const { Event } = require('../test_lib/event_decoder');
+
+const TokenHolder = artifacts.require('TokenHolder');
+
+contract('TokenHolder::revokeSession', async () => {
+    contract('Negative testing for input parameters:', async (accounts) => {
```
[the remark this answers is not in the document]
Comment (pgev) [reply test/token_holder/revoke_session.js]: As we have check that key should be authorized, this is already enough, as null can not be authorized.

File: test/token_holder/constructor.js [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+                    brandedTokenAddress,
+                    tokenRulesAddress,
+                    wallets,
+                    required,
+                ),
+                'Branded token address cannot be null.',
+            );
+        });
+
+        it('Null token rules address.', async () => {
```
Comment (abhayks1) [inline test/token_holder/constructor.js]: As discussed in previous PR, It is more readable if "it" makes a proper statement. It's easy to read and understand what test case is going to do. We can take others opinion also but this we have been following for quite long. Its fine if we want to take it up as part of improvements and update all at once.

Here for e.g. it('Fails for Null token rules address.') 
Your thoughts?

File: test/token_holder/revoke_session.js [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+            await utils.expectRevert(
+                tokenHolder.revokeSession(
+                    ephemeralKey,
+                    {
+                        from: registeredWallet0,
+                    },
+                ),
+                'Key to revoke is not in the authorized state.',
+            );
+        });
```
Comment (abhayks1) [inline test/token_holder/revoke_session.js]: One more use case for test:
Key revocation for expired ephemeral key should be success.

File: test/token_holder/submit_authorize_session.js [hunk 7]
```diff
@@ -0,0 +1,556 @@
...
+            await utils.expectRevert(
+                tokenHolder.submitAuthorizeSession(
+                    ephemeralKey,
+                    spendingLimit,
+                    expirationHeight,
+                    {
+                        from: registeredWallet0,
+                    },
+                ),
+                'Ephemeral key cannot be revoked.',
```
Comment (abhayks1) [inline test/token_holder/submit_authorize_session.js]: 'Ephemeral key cannot be revoked.' => 'Ephemeral key cannot be authorized.'

File: test/token_holder/submit_authorize_session.js [hunk 8]
```diff
@@ -0,0 +1,556 @@
...
+            await utils.expectRevert(
+                tokenHolder.submitAuthorizeSession(
+                    ephemeralKey,
+                    spendingLimit,
+                    expirationHeight,
+                    {
+                        from: registeredWallet0,
+                    },
+                ),
+                'Ephemeral key can not be expired.',
```
Comment (abhayks1) [inline test/token_holder/submit_authorize_session.js]: 'Ephemeral key can not be expired.' => 'Ephemeral key has expired.'

File: test/token_holder/submit_authorize_session.js [hunk 9]
```diff
@@ -0,0 +1,556 @@
...
+            await utils.expectRevert(
+                tokenHolder.submitAuthorizeSession(
+                    ephemeralKey,
+                    spendingLimit,
+                    (await web3.eth.getBlockNumber()),
+                    {
+                        from: registeredWallet0,
+                    },
+                ),
+                'Ephemeral key cannot be revoked.',
```
Comment (abhayks1) [inline test/token_holder/submit_authorize_session.js]: 'Ephemeral key cannot be revoked.' => 'Ephemeral key cannot be authorized.'

File: test/token_holder/submit_authorize_session.js [hunk 10]
```diff
@@ -0,0 +1,556 @@
...
+                    ephemeralKey, spendingLimit, expirationHeight,
+                ),
+            );
+
+            assert.isNotOk(
+                transaction.executed,
+            );
+        });
+    });
+});
```
Comment (abhayks1) [inline test/token_holder/submit_authorize_session.js]: More Test Case To Add:
- submitAuthorizeSession with spending limit = 0 should be success.

Comment (pgev) [reply test/token_holder/submit_authorize_session.js]: We don't need to check for all valid values. Only invalid ones.

**Comment by abhayks1** [comment]
LGTM.
Waiting for Travis to show green. I suggest to restart the build

**Pull request merged**